fms_extras/utils/cache/paged.py

```diff
@@ -0,0 +1,1044 @@
+    context_lens,
+    block_size,
+    max_context_len,
+    alibi_slopes=None,
+):
+    return out
+
+
+@torch.library.impl(lib, "paged_attention_v2", "CUDA")
+def _paged_attention_v2(
+    out,
+    exp_sums,
+    max_logits,
+    tmp_out,
+    query,
+    key_cache,
+    value_cache,
+    num_kv_heads,
+    scale,
+    block_tables,
+    context_lens,
+    block_size,
+    max_context_len,
+    alibi_slopes=None,
+):
+    attn_ops.paged_attention_v2(
+        out,
+        exp_sums,
+        max_logits,
+        tmp_out,
+        query,
+        key_cache,
+        value_cache,
+        num_kv_heads,
+        scale,
+        block_tables,
+        context_lens,
+        block_size,
+        max_context_len,
+        alibi_slopes,
+    )
+    return out
+
+
+lowering.fallbacks.add(torch.ops.paged_attention.paged_attention_v2)
+
+
+@lowering.register_lowering(
+    torch.ops.paged_attention.paged_attention_v2, type_promotion_kind=None
+)
+def _paged_attention_v2_lowering(
+    out,
+    exp_sums,
+    max_logits,
+    tmp_out,
+    query,
+    key_cache,
+    value_cache,
+    num_kv_heads,
+    scale,
+    block_tables,
+    context_lens,
+    block_size,
+    max_context_len,
+    alibi_slopes=None,
+):
+    PagedAttnKernel.create(
+        torch.ops.paged_attention.paged_attention_v2.default,
+        out,
+        exp_sums,
+        max_logits,
+        tmp_out,
+        query,
+        key_cache,
+        value_cache,
+        num_kv_heads,
+        scale,
+        block_tables,
+        context_lens,
+        block_size,
+        max_context_len,
+        alibi_slopes,
+        mutated_inputs=[out],
+    )
+    return out
+
+
+lib.define(
+    "paged_attention_v1(Tensor out, Tensor query, Tensor key_cache, Tensor value_cache, int num_kv_heads, float scale, Tensor block_tables, Tensor context_lens, int block_size, SymInt max_context_len, Tensor? alibi_slopes) -> Tensor"
+)
+
+
+@torch.library.impl(lib, "paged_attention_v1", "Meta")
+def _paged_attention_v1_meta(
+    out,
+    query,
+    key_cache,
+    value_cache,
+    num_kv_heads,
+    scale,
+    block_tables,
+    context_lens,
+    block_size,
+    max_context_len,
+    alibi_slopes=None,
+):
+    return out
+
+
+@torch.library.impl(lib, "paged_attention_v1", "CUDA")
+def _paged_attention_v1(
+    out,
+    query,
+    key_cache,
+    value_cache,
+    num_kv_heads,
+    scale,
+    block_tables,
+    context_lens,
+    block_size,
+    max_context_len,
+    alibi_slopes=None,
+):
+    attn_ops.paged_attention_v1(
+        out,
+        query,
+        key_cache,
+        value_cache,
+        num_kv_heads,
+        scale,
+        block_tables,
+        context_lens,
+        block_size,
+        max_context_len,
+        alibi_slopes,
+    )
+    return out
+
+
+lowering.fallbacks.add(torch.ops.paged_attention.paged_attention_v1)
+
+
+@lowering.register_lowering(
+    torch.ops.paged_attention.paged_attention_v1, type_promotion_kind=None
+)
+def _paged_attention_v1_lowering(
```

Comment:
We are doing same as vLLM for this call. Based on the comments in vLLM:

```
# NOTE(woosuk): We use a simple heuristic to decide whether to use
# PagedAttention V1 or V2. If the number of partitions is 1, we use
# V1 to avoid the overhead of reduction. Also, if the number of
# sequences or heads is large, we use V1 since there is enough work
# to parallelize.
# TODO(woosuk): Tune this heuristic.
# For context len > 8192, use V2 kernel to avoid shared memory shortage.
```